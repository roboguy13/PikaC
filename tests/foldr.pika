%generate foldrAdd

Sll : layout[-x](List);
Sll Nil := emp;
Sll (Cons head tail) :=
  exists +nxt.
  x :-> head **
  (x+1) :-> nxt **
  Sll tail [nxt];

foldrAdd : Int -> Sll -> Int;
foldrAdd z Nil := z;
foldrAdd z (Cons h t) := h + (foldrAdd z t);

-- cons : Int -> Sll -> Sll;
-- cons i Nil := Cons i (Nil [Sll]) [Sll];
-- cons i (Cons a as) := Cons i (Cons a as [Sll]) [Sll];
