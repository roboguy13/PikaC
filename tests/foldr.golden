void foldrAdd(loc i1, loc x, loc i2);
void foldrAdd(loc i1, loc x, loc i2) {
 if (x == 0) {
  WRITE_LOC(i2, 0, i1);
 } else {
  loc h = READ_LOC(x, 0);
  loc n = READ_LOC(x, 1);
  foldrAdd(i1, n, i2);
  loc i = READ_LOC(i2, 0);
  WRITE_LOC(i2, 0, (long)h + (long)i);
 }
}